packages/backend/src/models/JobModel/JobModel.ts

```diff
@@ -0,0 +1,175 @@
+            .orderBy('step_id', 'desc');
+
+        return steps.map((step) => ({
+            jobUuid: step.job_uuid,
+            createdAt: step.created_at,
+            updatedAt: step.updated_at,
+            stepStatus: step.step_status,
+            stepType: step.step_type,
+            stepLabel: step.step_label,
+        }));
+    }
+
+    async upsertJobStatus(
+        jobUuid: string,
+        projectUuid: string | undefined,
+        status: JobStatusType,
+    ): Promise<void> {
+        Logger.debug(
+            `Updating job status ${jobUuid} for project ${projectUuid} with status ${status}`,
+        );
+        await this.database(JobsTableName)
+            .insert({
+                project_uuid: projectUuid,
+                job_uuid: jobUuid,
+                job_status: status,
+                updated_at: new Date(),
+            })
+            .onConflict('job_uuid')
+            .merge();
+    }
+
+    async createJobSteps(
+        jobUuid: string,
+        stepTypes: JobStepType[],
+    ): Promise<void> {
+        await this.database.transaction(async (trx) => {
+            await stepTypes.reduce(async (previousPromise, step) => {
+                await previousPromise;
+                const stepLabel = JobLabels[step];
+                return trx(JobStepsTableName).insert({
+                    job_uuid: jobUuid,
+                    step_status: JobStepStatusType.PENDING,
+                    step_type: step,
+                    step_label: stepLabel,
+                });
+            }, Promise.resolve());
+        });
+    }
+
+    async updateJobStep(
+        jobUuid: string,
+        stepStatus: JobStepStatusType,
+        stepType: JobStepType,
+    ): Promise<void> {
+        await this.database(JobStepsTableName)
+            .update({
+                step_status: stepStatus,
+                updated_at: new Date(),
+            })
+            .where('job_uuid', jobUuid)
+            .andWhere('step_type', stepType);
+
+        await this.database(JobsTableName)
+            .update({
+                updated_at: new Date(),
+                job_status:
+                    stepStatus === JobStepStatusType.ERROR
+                        ? JobStatusType.ERROR
+                        : JobStatusType.RUNNING,
+            })
+            .where('job_uuid', jobUuid);
+    }
+
+    async tryJobStep<T>(
+        jobUuid: string,
+        jobStepType: JobStepType,
+        callback: () => Promise<T>,
+    ): Promise<T> {
+        try {
+            await this.updateJobStep(
+                jobUuid,
+                JobStepStatusType.RUNNING,
+                jobStepType,
+            );
+
+            const result = await callback();
+
+            await this.updateJobStep(
+                jobUuid,
+                JobStepStatusType.DONE,
+                jobStepType,
+            );
+            return result;
+        } catch (e) {
+            await this.updateJobStep(
```

Comment:
If we store the error message in the job, then the whole thing can complete async, and we'd discover the error by checking the job. wdyt?

@ZeRego @rephus we didn't actually discuss what would happen on error but I assumed that's how it would work

Reply:
I think we would only store errors from the `callback`